cmd/login/wc.go

```diff
@@ -147,65 +200,90 @@ func (r *runner) createClusterClientCert(ctx context.Context, client k8sclient.I
 		certOperatorVersion: certOperatorVersion,
 	}
 
-	clientCertResource, err := createCert(ctx, clientCertService, certConfig)
+	clientCertResource, secret, err := r.getCredentials(ctx, services.clientCertService, certConfig)
 	if err != nil {
-		return microerror.Mask(err)
+		return "", false, microerror.Mask(err)
 	}
 
-	// Retrieve client certificate credential.
-	secret, err := fetchCredential(ctx, provider, clientCertService, clientCertResource)
+	credentialConfig := credentialConfig{
+		clusterID:     r.flag.WCName,
+		certCRT:       secret.Data[credentialKeyCertCRT],
+		certKey:       secret.Data[credentialKeyCertKey],
+		certCA:        secret.Data[credentialKeyCertCA],
+		clusterServer: getServerAddress(certConfig),
+		filePath:      r.flag.SelfContained,
+		loginOptions:  r.loginOptions,
+	}
+
+	contextName, contextExists, err := r.storeWCCredentials(credentialConfig)
 	if err != nil {
-		return microerror.Mask(err)
+		return "", false, microerror.Mask(err)
 	}
 
-	// Store client certificate credential either into the current kubeconfig or a self-contained file if a path is given.
-	var contextExists bool
-	var contextName string
-	if r.loginOptions.selfContainedWC {
-		contextName, contextExists, err = printWCCredentials(r.k8sConfigAccess, r.fs, r.flag.SelfContained, clientCertResource, secret, clusterBasePath, r.loginOptions)
+	if clientCertResource != nil {
+		// Cleaning up leftover resources.
+		err = cleanUpClientCertResources(ctx, services.clientCertService, clientCertResource)
 		if err != nil {
-			return microerror.Mask(err)
+			return "", false, microerror.Mask(err)
 		}
-	} else {
-		contextName, contextExists, err = storeWCCredentials(r.k8sConfigAccess, r.fs, clientCertResource, secret, clusterBasePath, r.loginOptions)
+	}
+
+	fmt.Fprint(r.stdout, color.GreenString("\nCreated client certificate for workload cluster '%s'.\n", r.flag.WCName))
+	return contextName, contextExists, nil
+}
+
+func (r *runner) getCredentials(ctx context.Context, clientCertService clientcert.Interface, config clientCertConfig) (*clientcert.ClientCert, *v1.Secret, error) {
+	var secret *v1.Secret
+	var clientCert *clientcert.ClientCert
+	var err error
+
+	if config.provider == key.ProviderOpenStack {
+		// Retrieve the WC CA-secret.
+		ca, err := clientCertService.GetCredential(ctx, config.clusterNamespace, config.clusterName+"-ca")
 		if err != nil {
-			return microerror.Mask(err)
+			return nil, nil, microerror.Mask(err)
 		}
+		secret, err = generateCredential(ctx, ca, config)
+		if err != nil {
+			return nil, nil, microerror.Mask(err)
+		}
+		return nil, secret, nil
```

Comment:
Why don't you generate the client cert here as you do below for other providers?

Reply:
The function that is used for other providers does not directly create the client certificate. It creates a `certConfig` which is applied to the MC. `cert-operator` reconciles it and then creates the certificate which is fetched. Later we need to clean up the `certConfig` that was created. (that is why we return it)
Since `cert-operator` is not part of CAPO, this does not work out of the box. We could do something similar, but I don't think this approach of creating something on the MC aligns with the gitops principle.
Let me know if there are better options, though.